Fix/column width units

tellthemachines opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/master/docs/contributors/repository-management.md#pull-requests. -->

## Description
<!-- Please describe what you have changed or added -->

Fixes #26096 and #26641

#24711 introduced two issues when column width attributes switched from being defined as numbers to strings:

* Some of the utility functions for calculating width when changing column number broke as they expected to always receive numbers;
* Templates defining widths as numbers broke as they expected a number to always be interpreted as a percent.

This PR:
* changes two of the utility functions to parse strings for numbers and use them as such; 
*  re-introduces the previous behaviour of  adding a `%` to the end of a column width when it is defined as a number. 

To avoid breaking columns already set with fixed widths, recalculation on adding/removing columns only takes into account the existing column widths if they are defined in percent, or if they are unitless (which should be interpreted as percent for back compat reasons).

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

Tested in browser:
 * Add a 30/70 columns block, add a new column to it and verify new widths look as expected. 
* Add/remove columns to various blocks with explicit widths defined. Check that widths defined in `px` don't get changed when adding or removing columns.
* Check that unitless widths in a custom post type template are interpreted as `%`.

Also added a few unit tests for the changes in the utility functions.

## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
 Bug fix (non-breaking change which fixes an issue)
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

tellthemachines committed:
Fix issues with different units in column widths

tellthemachines committed:
Columns with fixed width should keep width on recalculation

ntsekouras commented on packages/block-library/src/columns/utils.js:
```diff
@@ -96,10 +99,18 @@ export function getRedistributedColumnWidths(
-		Number.isFinite( block.attributes.width )
-	);
+export function hasExplicitPercentColumnWidths( blocks ) {
+	return blocks.every( ( block ) => {
```
I think you can simplify this a bit like:
```
const blockWidth = block.attributes.width;
return Number.isFinite(
	blockWidth.endsWith?.( '%' ) ? parseFloat( blockWidth ) : blockWidth
);
```

tellthemachines replied:
Ooooh, much better, thanks!

ntsekouras commented on packages/block-library/src/column/deprecated.js:
```diff
@@ -6,9 +6,34 @@ import classnames from 'classnames';
+import { InnerBlocks, useBlockProps } from '@wordpress/block-editor';
 
 const deprecated = [
+	{
```
I'm not really sure about the deprecation here. For start it should have `attributes` or `isEligible` etc.. With just a `save` function, wouldn't it match if a Columns block was invalid for whatever reason? -cc @gziolo

gziolo replied:
It would work without `attributes`, but as soon as they change, it might stop working. So it's safer to set them. `isEligible` is optional.

ntsekouras reviewed (commented):
Hey @tellthemachines - thanks for working on this. I tested it and seems fine! I'm not really sure about the deprecation though about what cases wants to handle..

gziolo commented on packages/block-library/src/column/deprecated.js:
```diff
@@ -6,9 +6,34 @@ import classnames from 'classnames';
+			} );
+
+			let style;
+			if ( width ) {
```
Is it possible to modify the previous deprecation instead? The one that created the issue in the first place?

gziolo replied:
If we have https://github.com/WordPress/gutenberg/pull/26757/files#r518737391 + https://github.com/WordPress/gutenberg/pull/26774, then we might not need another deprecation.

gziolo commented:
It would be great to add some tests that verify if the originally introduced deprecation work as expected in the first place. We should also add a block fixture that is buggy to ensure it's resolved.

gziolo commented:
I opened #26774 that adds tests for the deprecation introduced in #24711. It also improves a bit the setup for the deprecation based on the suggestions from @ntsekouras. It would be nice to land first and rebase this PR to ensure this test still passes.

gziolo commented on packages/block-library/src/column/save.js:
```diff
@@ -16,7 +16,10 @@ export default function save( { attributes } ) {
 	let style;
-	if ( width ) {
+
+	if ( Number.isFinite( width ) ) {
```
`width` can be undefined, so we still should keep the check, how about (I included code comment to ensure it doesn't get removed:

```js
if ( width ) {
	// Numbers are handled for backward compatibility when provided with templates.
	style = { flexBasis: Number.isFinite( width ) ? width + '%' : width };
}
```

tellthemachines committed:
Address review feedback

tellthemachines commented:
I addressed all the feedback, removed the deprecation and ran the integration tests locally; everything looks fine. Also tested with a few column blocks created on a previous branch and nothing breaks in the editor. 

E2e failures don't seem related to this work; they're failing on every PR now 😩 

Would be good to get this merged today though, so we can include it in Beta 4.

aristath reviewed (approved):
LGTM :+1:

gziolo commented on packages/block-library/src/column/save.js:
```diff
@@ -16,8 +16,10 @@ export default function save( { attributes } ) {
 	if ( width ) {
-		style = { flexBasis: width };
+		// Numbers are handled for backward compatibility as they can be still provided with templates.
+		style = { flexBasis: Number.isFinite( width ) ? width + '%' : width };
```
I will explore it further, but seeing this PR again, I started thinking about whether new changes applied to `edit` are enough. Anyway, I can send a follow-up PR when I confirm it isn't necessary anymore. For now, better safe than sorry 😄

gziolo replied:
I tested it with the following diff:

```diff
diff --git a/packages/block-library/src/column/save.js b/packages/block-library/src/column/save.js
index c43edc8464..ccf1b64387 100644
--- a/packages/block-library/src/column/save.js
+++ b/packages/block-library/src/column/save.js
@@ -18,8 +18,7 @@ export default function save( { attributes } ) {
 	let style;
 
 	if ( width ) {
-		// Numbers are handled for backward compatibility as they can be still provided with templates.
-		style = { flexBasis: Number.isFinite( width ) ? width + '%' : width };
+		style = { flexBasis: width };
 	}
 
 	return (
diff --git a/packages/block-library/src/columns/variations.js b/packages/block-library/src/columns/variations.js
index 9f20699dee..6026a5f51e 100644
--- a/packages/block-library/src/columns/variations.js
+++ b/packages/block-library/src/columns/variations.js
@@ -74,8 +74,8 @@ const variations = [
 			</SVG>
 		),
 		innerBlocks: [
-			[ 'core/column', { width: '33.33%' } ],
-			[ 'core/column', { width: '66.66%' } ],
+			[ 'core/column', { width: 33.33 } ],
+			[ 'core/column', { width: 66.66 } ],
 		],
 		scope: [ 'block' ],
 	},
```

and it seems to work correctly. Do you think we should update codebase or do I miss some other case?

tellthemachines replied:
Tested this by adding a 30/70 columns block, and both columns are rendering with the same size. But that happens with or without your change to `save.js`, so looking for the issue now 🤔

tellthemachines replied:
Scratch that, I was passing the template widths in as unitless strings 😅 . That's not something we'd ever expect to happen, is it?

I think it's fine to make this change, anyway!

gziolo replied:
Yes, it works properly only with numbers. I had the same issue with strings. We can always check the issue filed and previous setup before it was converted to percentages. If it worked with strings in the past, we might need another fix 😃

gziolo reviewed (approved):
Looks good 🚢

lukewalczak reviewed (approved):
Tested changes on mobile and didn't spot any breakage 🎉
